src/triage/component/catwalk/storage.py

```diff
@@ -339,30 +340,101 @@ def __init__(
+
+        Returns both as a tuple
+        """
+        labels = matrix_with_labels.pop(self.label_column_name)
+        # After we pop the labels from the matrix with labels it is now a design matrix
+        design_matrix = matrix_with_labels
+        if self.should_cache:
+            # we neither want to store multiple copies of the matrix
+            # nor misrepresent the label-less version of the matrix as the one with labels
+            # so we explicitly set the matrix_with_labels to None in the cache
+            self.matrix_with_labels = None
+            self.design_matrix = design_matrix
+            self.labels = labels
+        return design_matrix, labels
 
     @property
-    def matrix(self):
-        """The raw matrix. Will load from storage into memory if not already loaded"""
-        if self.__matrix is None:
-            self.__matrix = self._load()
-            # Is the index already in place?
-            if self.__matrix.index.names != self.metadata['indices']:
-                self.__matrix.set_index(self.metadata['indices'], inplace=True)
+    def design_matrix(self):
+        """The matrix without the label vector, only the index and features"""
+        if self.__design_matrix is not None:
+            return self.__design_matrix
+        else:
+            return self._split_matrix(self.matrix_with_labels)[0]
 
-            self.__matrix = downcast_matrix(self.__matrix)
-        return self.__matrix
+    @design_matrix.setter
+    def design_matrix(self, design_matrix):
+        self.__design_matrix = design_matrix
 
-    @matrix.setter
-    def matrix(self, matrix):
-        self.__matrix = matrix
+    @property
+    def labels(self):
+        """The label vector of the matrix"""
+        if self.__labels is not None:
+            return self.__labels
+        return self._split_matrix(self.matrix_with_labels)[1]
+
+    @labels.setter
+    def labels(self, labels):
+        self.__labels = labels
+
+    @property
+    def matrix_with_labels(self):
+        """The matrix complete with label vector."""
+        if self.__matrix_with_labels is not None:
+            return self._preprocess_matrix_with_labels(self.__matrix_with_labels)
+        elif self.__design_matrix is not None and self.__labels is not None:
+            return self.design_matrix.assign(**{self.label_column_name: self.labels})
+        else:
+            return self._preprocess_matrix_with_labels(self._load())
+
+    @matrix_with_labels.setter
+    def matrix_with_labels(self, matrix_with_labels):
+        self.__matrix_with_labels = matrix_with_labels
 
     @property
     def metadata(self):
         """The raw metadata. Will load from storage into memory if not already loaded"""
-        if self.__metadata is None:
-            self.__metadata = self.load_metadata()
+        if self.__metadata is not None:
+            return self.__metadata
+        if not self.should_cache or self.__metadata is None:
```

Comment:
What does `should_cache` have to do with this condition? I would think that if you don't have the data, you have to load it, regardless.

Reply:
Fair enough. I think that was left over from before there was the first shortcut `if`.